go/marshal/encode_type_test.go

```diff
@@ -0,0 +1,377 @@
+	assert.NoError(err)
+	assert.True(types.MakeMapType(
+		types.MakeStructTypeFromFields("S", types.FieldMap{
+			"n": types.StringType,
+		}),
+		types.BoolType).Equals(typ))
+}
+
+func TestMarshalTypeSet(t *testing.T) {
+	assert := assert.New(t)
+
+	type S struct {
+		A map[int]struct{} `noms:",set"`
+		B map[int]struct{}
+		C map[int]string      `noms:",set"`
+		D map[string]struct{} `noms:",set"`
+		E map[string]struct{}
+		F map[string]int `noms:",set"`
+		G []int          `noms:",set"`
+		H string         `noms:",set"`
+	}
+	var s S
+	typ, err := MarshalType(s)
+	assert.NoError(err)
+
+	emptyStructType := types.MakeStructTypeFromFields("", types.FieldMap{})
+
+	assert.True(types.MakeStructTypeFromFields("S", types.FieldMap{
+		"a": types.MakeSetType(types.NumberType),
+		"b": types.MakeMapType(types.NumberType, emptyStructType),
+		"c": types.MakeMapType(types.NumberType, types.StringType),
+		"d": types.MakeSetType(types.StringType),
+		"e": types.MakeMapType(types.StringType, emptyStructType),
+		"f": types.MakeMapType(types.StringType, types.NumberType),
+		"g": types.MakeListType(types.NumberType),
+		"h": types.StringType,
+	}).Equals(typ))
+}
+
+func TestMarshalTypeSetWithTags(t *testing.T) {
+	assert := assert.New(t)
+
+	type S struct {
+		A map[int]struct{} `noms:"foo,set"`
+		B map[int]struct{} `noms:",omitempty,set"`
+		C map[int]struct{} `noms:"bar,omitempty,set"`
+	}
+
+	var s S
+	typ, err := MarshalType(s)
+	assert.NoError(err)
+	assert.True(types.MakeStructTypeFromFields("S", types.FieldMap{
+		"foo": types.MakeSetType(types.NumberType),
+		"b":   types.MakeSetType(types.NumberType),
+		"bar": types.MakeSetType(types.NumberType),
+	}).Equals(typ))
+}
+
+func TestMarshalTypeInvalidTag(t *testing.T) {
+	type S struct {
+		F string `noms:",omitEmpty"`
+	}
+	var s S
+	_, err := MarshalType(s)
+	assert.Error(t, err)
+	assert.Equal(t, `Unrecognized tag: omitEmpty`, err.Error())
+}
+
+func TestMarshalTypeCanSkipUnexportedField(t *testing.T) {
+	assert := assert.New(t)
+
+	type S struct {
+		Abc         int
+		notExported bool `noms:"-"`
+	}
+	var s S
+	typ, err := MarshalType(s)
+	assert.NoError(err)
+	assert.True(types.MakeStructTypeFromFields("S", types.FieldMap{
+		"abc": types.NumberType,
+	}).Equals(typ))
+}
+
+func TestMarshalTypeOriginal(t *testing.T) {
+	type S struct {
+		Foo int          `noms:",omitempty"`
+		Bar types.Struct `noms:",original"`
+	}
+
+	var s S
+	assertMarshalTypeErrorMessage(t, s, "Type is not supported, type: marshal.S")
+
+}
+
+func TestMarshalTypeNomsTypes(t *testing.T) {
+	assert := assert.New(t)
+
+	type S struct {
+		Blob   types.Blob
+		Bool   types.Bool
+		Number types.Number
+		String types.String
+		Type   *types.Type
+	}
+	var s S
+	assert.True(MustMarshalType(s).Equals(
+		types.MakeStructTypeFromFields("S", types.FieldMap{
+			"blob":   types.BlobType,
+			"bool":   types.BoolType,
+			"number": types.NumberType,
+			"string": types.StringType,
+			"type":   types.TypeType,
```

Comment:
None of the noms types that take type parameters are supported. I'll add tests that these errors